proposals/001-operator-design.md

```diff
@@ -0,0 +1,818 @@
+
+### 2. Off-Cluster access
+
+```yaml
+---
+kind: KafkaGateway
+apiVersion: kroxylicious.io/v1alpha1
+metadata:
+  name: simple
+  namespace: my-namespace
+spec:
+  listeners:
+    - name: my-listener
+      port: 9092 # required
+      protocol: kroxylicious.io/kafka-tls
+      tls:
+        certificateRefs:
+          - # user supplied secret containing tls.crt and tls.key
+            kind: Secret
+            name: my-cert
+            namespace: my-namespace
+---
+kind: KafkaRoute
+apiVersion: kroxylicious.io/v1alpha1
+metadata:
+  name: my-route
+  namespace: my-namespace
+spec:
+  parentRefs:
+    - group: kroxylicious.io
+      kind: KafkaGateway
+      name: simple
+      sectionName: my-listener
+      brokers:
+        - advertisedBrokerIds: [ 1,2,3 ]
+          brokerIdentityMapping: identity
+      hostnames:
+        - my-cluster-%.example.com
+      rules:
+        - name: my-rule
+          filters: [ ]
+          backendRefs:
+            - kind: Service
+              namespace: kafka
+              name: my-cluster
+              port: 9093
+```
+
+Because the hostname is not suffixed with `svc.cluster.local` the operator will infer that we want to expose this
+hostname off-cluster.
+
+The operator will create a LoadBalancer Service named `my-route` mapping port 9092 to all the Proxy instances
+
+The proxy will be configured with a single VirtualCluster with SNI exposition (
+bootstrap == `my-cluster-bootstrap.example.com`, brokerPattern = `my-cluster-$(nodeId).example.com`).
+
+The VirtualCluster will have the certificate wired in and know how to select the certificate as discussed here.
+
+The virtual cluster target cluster bootstrap will be configured as `my-cluster.kafka:9093`
+
+The user will be responsible for configuring external DNS to route those hostnames to the loadbalancer IP and they can
+bootstrap with `my-cluster-bootstrap.example.com:9092`
+
+### 3. SNI external DNS using OpenShift Routes
+
+With this solution we would leverage OpenShift Routes for getting the traffic to the Proxy. It would work the same as it
+does for [Strimzi](https://strimzi.io/blog/2019/04/30/accessing-kafka-part-3/).
+
+The Kroxylicious Operator would need to create Routes (passthrough) for each bootstrap and broker endpoint. This would
+be in addition to the per bootstrap/broker Kube Service The OpenShift’s Ingress Operator automatically sets up external
+DNS so that those names resolve externally.
+
+Advantages:
+
+- DNS twiddling is automated by the platform.
+
+Disadvantage:
+
+- It’s OpenShift only
+- We're proxying the proxy with HAProxy. More latency. More machinery.
+- Known issues with haproxy getting restarted every time a new Route is created = dropped Kafka connections. However
+  they seem to be fixing that (https://issues.redhat.com/browse/NE-879)!
+
+We could have a configurable list of ingress domains in KafkaGatewayParameters, then when we find a hostname that is a
```

Comment:
It is somewhat unfortunate to need to duplicate the ingressdomain.  If the ingressdomain changes for whatever, the configuration in the Kroxylicious CRs would be stale.